### PR TITLE
Improve error handling in the `sighash` module

### PR DESCRIPTION
**Pull request by tcharding**
Improve the error handling in the `sighash` module by adding small specific error types. 

Close: #2150

### CODE REVIEW

**Review by apoelstra — APPROVED** [review]
ACK 665ae728b90f01291da9eee8feb84044b42733a0

File: bitcoin/src/blockdata/transaction.rs
Comment (Kixunil) [inline bitcoin/src/blockdata/transaction.rs:991]: A note about this being better than indexing by providing `Result` would be nice.

File: bitcoin/src/blockdata/transaction.rs
Comment (Kixunil) [inline bitcoin/src/blockdata/transaction.rs:1024]: This is fine, especially with `non_exhaustive` and I think we can also commit to these fields but it might be really cool to also provide txid so that if one runs a program processing many transactions the problematic one it can be easily found.

However hashing may be a bit expensive even for error path, so maybe this should be behind `fancy_errors` feature or something.

Also we could carry the information whether it was inputs or outputs.

File: bitcoin/src/crypto/sighash.rs
Comment (Kixunil) [inline bitcoin/src/crypto/sighash.rs:1294]: Well, then this should be an enum differentiating between these error cases. But it's fine for now.

File: bitcoin/src/crypto/sighash.rs
Comment (Kixunil) [inline bitcoin/src/crypto/sighash.rs:1354]: Why is this `ErrorKind` and not `Error`?

File: bitcoin/src/psbt/mod.rs
Comment (Kixunil) [inline bitcoin/src/psbt/mod.rs:773]: Oh, so why `sighash::Error` doesn't contain both `TaprootError` and `P2wpkhError`?

File: bitcoin/src/crypto/sighash.rs
Comment (Kixunil) [inline bitcoin/src/crypto/sighash.rs:1352]: This one is in `TaprootError` and is never in `P2wpkhError` so it should not be here?

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs:1354]: I wondered the same when I was doing this, I don't remember when this got changed/created.

**Comment by tcharding** [comment]
Thanks for the review @Kixunil, I'll go back over the whole PR.

File: bitcoin/src/crypto/sighash.rs
Comment (apoelstra) [reply bitcoin/src/crypto/sighash.rs:1354]: In a commit message for https://github.com/rust-bitcoin/rust-bitcoin/pull/2066:

>    Sadly the `std::io::Error` type requires `std::error::Error` as a
>    bound on the inner error, which is rather difficult to duplicate in
>    a way that allows for mapping to `std` and back.

Comment (Kixunil) [reply bitcoin/src/crypto/sighash.rs:1354]: But why not our `io::Error`? We do have the type. I just tried it locally and it works if we drop the derives on errors.

Comment (apoelstra) [reply bitcoin/src/crypto/sighash.rs:1354]: If it works we should switch to `io::Error` ... #2066 predates the `io` crate and is from an era when we were trying to glue core2 and std together in a transparent way, and it may be that we can do things now that we didn't use to be able to.

We use `io::ErrorKind` in a few places in this codebase. May be worth auditing those.

File: bitcoin/src/crypto/sighash.rs
Comment (Kixunil) [inline bitcoin/src/crypto/sighash.rs:1227]: It'd be useful to make this a separate type too for my PR. I'll push mine and we'll resolve who does what.

Also I've also created `TaprootError` but a bit different.

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs:1352]: Yeah it is returned when getting the txin `let txin = self.tx.borrow().txin(input_index)?;`

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs]: Done, see `SingleMissingOutputError`.

File: bitcoin/src/psbt/mod.rs
Comment (tcharding) [reply bitcoin/src/psbt/mod.rs:773]: Please see new version, no includes the following:
- `TaprootError`
- `SegwitV0Error`
- `LegacyError`

Eventhough the bottom two have the same fields, I believe this clears up the confusion in `psbt`.

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs]: Improved, done as a separate patch.

File: bitcoin/src/blockdata/transaction.rs
Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs:1030]: I don't want to introduce a new feature in this PR, I agree we should consider having one but there that smells like there will be a bit of back-and-forth on that.

I'm not totally happy with the.

File: bitcoin/src/blockdata/transaction.rs
Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs:997]: Done.

**Comment by tcharding** [comment]
Notes on new version:
- I have not touched the `io::ErrorKind` vs `io::Error` issue, it is orthogonal to this PR
- I have not added the additional information to `IndexOutOfBoundsError` (input vs output vec, and txid). I request that this is done as a follow up.

All other review suggestions seen to I believe.

File: bitcoin/src/psbt/mod.rs
Comment (Kixunil) [inline bitcoin/src/psbt/mod.rs:771]: This variant looks redundant since it's contained in the other one? Though this gives information that the operation didn't involve P2WPKH...

**Review by Kixunil — COMMENTED** [review]
ACK 2904c7a13a86e9052f02d77a33f1063d062b5ed8 but I really want the message to be fixe in a followup PR

**Review by Kixunil — DISMISSED** [review]

File: bitcoin/src/blockdata/transaction.rs
Comment (apoelstra) [inline bitcoin/src/blockdata/transaction.rs:999]: In 440dee61ad52cbcf0818d0d1a862abb585827c48:

Should be `tx_in` (and `tx_out`) to match capitalization of `TxIn`/`TxOut`. (Ok to do this in a separate commit in case you don't want to rebase your whole commit stack.)

**Comment by apoelstra** [comment]
ACK 2904c7a13a86e9052f02d77a33f1063d062b5ed8 except the naming thing which I think is blocking. (And then you might as well address Kix's nits :))

File: bitcoin/src/blockdata/transaction.rs
Comment (apoelstra) [reply bitcoin/src/blockdata/transaction.rs:1034]: I'd also like the bool in there.

File: bitcoin/src/blockdata/transaction.rs
Comment (apoelstra) [reply bitcoin/src/blockdata/transaction.rs:1040]: Are you saying we should leave it out (using the default impl) or that we shouldn't implement `std::Error` at all?

I think we have a soft policy (unsure if it's written down) to never use the default impl, specifically so that with enums we'll get compilation errors when we add variants and won't forget. I guess it's reasonable to say that this rule doesn't apply to structs.

File: bitcoin/src/crypto/sighash.rs
Comment (apoelstra) [reply bitcoin/src/crypto/sighash.rs:1338]: Nice catch!! This is the kind of thing I never notice until months later.

File: bitcoin/src/crypto/sighash.rs
Comment (apoelstra) [reply bitcoin/src/crypto/sighash.rs:1019]: It would also demonstrate the value of having these fine-grained error types at all :)

File: bitcoin/src/blockdata/transaction.rs
Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs:1034]: oooo, yep I agree that is a decent rule violation. I toyed with having a bool, also with having an additional error type `InputIndexOutOfBounds`, that led me to wonder do we _really_ need the `txout` function, and that led me to throw my hands up and just leave it as it was. No excuse for the incorrect error message though, that was pure mistake.

Ok, I found a place where we can use `txout` (in `sighash` module), I'm going to add two errors, one for inputs and one for outputs.

File: bitcoin/src/blockdata/transaction.rs
Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs:1040]: Yeah I also thought we had a soft policy, I was thinking that being explicit at the cost of a single line makes the code more readable for newer devs, especially application developers who mightn't be as familiar with error boilerplate. For structs I'm not bothered either way, but I'd like a decision so we can be uniform.

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs:1326]: Sure thing, will change.

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs:1328]: Sure, can do.

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs:1338]: Yes, nice one.

File: bitcoin/src/blockdata/transaction.rs
Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs:999]: Rebasing is part of my job ;) This is tongue in cheek but I don't understand what is the difficulty with rebasing, I rebase my patch sets constantly all day and move changes up and down the patch set into the right place. I would also rebase on master constantly but I resist so that reviewers can do `git range-diff` and get less cluttered output.

Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs:999]: But yes, `tx_in` and `tx_out` are better names. Sorry I'm a bit prickly this week aren't I.

Comment (apoelstra) [reply bitcoin/src/blockdata/transaction.rs:999]: All good :)

And usually rebases aren't too big a concern, but renaming a constant that's used in a bunch of future patches can be pretty annoying and lead to tons of conflicts.

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs:1337]: Used as suggested.

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs:1157]: Made all three prevouts error variants more terse since they have descriptive inner error types.

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs:1275]: Used as suggested.

File: bitcoin/src/psbt/mod.rs
Comment (tcharding) [reply bitcoin/src/psbt/mod.rs:771]: Fair point, it originates in a different error path. But more to the point, I'm not fused about `psbt` stuff here. I'll keep it in mind over in `tcharding/rust-psbt` though when I upgrade to this version of `rust-bitcoin`.

File: bitcoin/src/blockdata/transaction.rs
Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs]: Well, you may be amused to hear that after being so facetious I had to rebase this PR about 10 times to get everything in the right place :)

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs:1380]: Soon as `io` is released hopefully we will get a few nice clean ups and improvements around the `io::Error`.

**Comment by tcharding** [comment]
All review suggestions should be either seen to or commented on. Thanks

File: bitcoin/src/crypto/sighash.rs
Comment (Kixunil) [reply bitcoin/src/crypto/sighash.rs:1381]: I don't think this needs to be blocked on IO release. Just a followup PR. If this one merges before your morning I'll likely do it myself.

File: bitcoin/src/blockdata/transaction.rs
Comment (Kixunil) [reply bitcoin/src/blockdata/transaction.rs:1072]: Leave out `source`.

> specifically so that with enums we'll get compilation errors when we add variants and won't forget. I guess it's reasonable to say that this rule doesn't apply to structs.

Exactly this.

> more readable for newer devs, especially application developers who mightn't be as familiar with error

I don't see this as worth it and it may actually confuse the developers ("why is is this reimplementing the same thing?"). Also in the past needlessly overriding defaults caused some other problem (I don't remember what it was).

I might consider it in some cases if the impl was `let &Self { field1: _, field2: _ } = self; None` so it prevents forgetting to update it if we add the source (or change to enum). This case of a clearly leaf error isn't one of them.

File: bitcoin/src/blockdata/transaction.rs
Comment (Kixunil) [inline bitcoin/src/blockdata/transaction.rs:1041]: Either this or the `source` implementation is wrong - in `std` it will never display the inner information.

You should use `write` and directly access the fields of the inner error to improve the error message because just switching to `write!` would create this long redundant message:

> invalid output index: index {} is out-of-bounds for vector with length {}

But anyway, I think that this approach is either confusing or creates needlessly long message if you return `source`, so I'm not really entirely happy about it (but it may be the least bad option, didn't see the rest of the code yet).

File: bitcoin/src/blockdata/transaction.rs
Comment (Kixunil) [inline bitcoin/src/blockdata/transaction.rs:1037]: There's a question whether this should be a named field since `#[non_exhaustive]` suggest we could add more fields but we'd have to keep it as tuple struct which wouldn't be great. We could remove `non_exhaustive` but there's a possible different choice: make `IndexOutOfBounds` error general-purpose, moving it from `transaction` module, so things like `txid` won't get added into that one but into this one. This is somewhat nice except for long error message.

File: bitcoin/src/crypto/sighash.rs
Comment (Kixunil) [inline bitcoin/src/crypto/sighash.rs:1326]: "initial prefix" sounds redundant.

File: bitcoin/src/crypto/sighash.rs
Comment (Kixunil) [inline bitcoin/src/crypto/sighash.rs:1150]: Here's an idea, let's have:

```rust
pub enum Error<E> {
    Io(io::Error),
    Sighash(E),
}

impl<E> Error<E> {
    fn unwrap_sighash(self) -> E {
        match self {
            Self::Sighash(error) => error,
            Self::Io(error) => panic!("hash engine error {}", error),
        }
    }
}
```

Then we can retain the derives on this type even when it's `io::Error` and call `unwrap_sighash` in methods returning hashes. It also avoids duplication of the `Io` variant across all enums.

**Review by Kixunil — CHANGES_REQUESTED** [review]
NACK because of broken `Error` impl but I'd really, really like to have the `Error` enum (see below) too.

File: bitcoin/src/blockdata/transaction.rs
Comment (apoelstra) [reply bitcoin/src/blockdata/transaction.rs:1007]: I think `input` and `output` are confusing. A "txout" or "txin" is a noun by itself. If we rename these then we're gratuitously using different terminology than everyone else.

But I don't care too much about this, so I'll concede if you feel strongly.

Comment (Kixunil) [reply bitcoin/src/blockdata/transaction.rs:1007]: What is confusing about them?

File: bitcoin/src/blockdata/transaction.rs
Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs:1037]: Oh I did not realize that we couldn't use `non_exhaustive` with tuple structs as you describe. There are a bunch of these in the code base now. Ugh `git grep -B 2 'pub struct .*Error'` shows we have non-uniform usage of `pub` on the inner filed as well.

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs:1326]: Ha! You are correct.

File: bitcoin/src/crypto/sighash.rs
Comment (tcharding) [reply bitcoin/src/crypto/sighash.rs:1150]: At first blush, that looks like a good idea.

File: bitcoin/src/blockdata/transaction.rs
Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs:1007]: Damn, naming is hard. I agree with @Kixunil, from a purely Rust perspective we should have (ignoring other fields)
```rust
struct Transaction {
    inputs: Vec<TxIn>
}

impl Transaction {
    pub fn input(&self, index) -> Result<&TxIn, Error> { .... }
}
```

But I tend to agree with @apoelstra that this is not very bitcoin-y

File: bitcoin/src/blockdata/transaction.rs
Comment (Kixunil) [reply bitcoin/src/blockdata/transaction.rs:1037]: > we have non-uniform usage of `pub` on the inner filed as well.

That's fine, we don't have to commit to everything and should decide case-by-case.

File: bitcoin/src/blockdata/transaction.rs
Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs:1041]: Excluding that I was very much wrong returning `None` from `source`, the error message is as I expected it to be. What is confusing about it?
E.g with numbers:
`invalid output index: index 2 is out-of-bounds for vector with length 1`

File: bitcoin/src/blockdata/transaction.rs
Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs:1037]: Ok, this PR is holding up other work. Can we just put it in like this? There are a bunch of these tuple struct error types in the codebase now with `non_exhaustive` on them that all have this problem. They all need fixing, adding this one is not making the code materially worse.

Comment (tcharding) [reply bitcoin/src/blockdata/transaction.rs:1037]: Flagging this thread: Needs explicit ack from @Kixunil before merging.

**Comment by tcharding** [comment]
I request to merge this as is, not the following things need doing as follow ups:
- Resolve the naming of `Transaction::tx_in` and `tx_out`
- Look into the `ErrorKind` issue and solution described above
- Look into the usage of `non_exhaustive` on tuple structs


Note also, please don't merge without ack for Kix (mentioned in thread above already).

File: bitcoin/src/blockdata/transaction.rs
Comment (Kixunil) [reply bitcoin/src/blockdata/transaction.rs:1041]: There are two options to impl `Error`: return `Some` and impl display, which produces long messages, or delegate everything to the inner error. The latter is confusing in that if you do non-trivial inspecting of `source()` somewhere up the chain this will skip the wrapped type, so you can't downcast to it. I guess long error messages might be the lesser evil.

File: bitcoin/src/blockdata/transaction.rs
Comment (Kixunil) [reply bitcoin/src/blockdata/transaction.rs:1037]: Yes, but needs a 1.0 issue.

File: bitcoin/src/blockdata/transaction.rs
Comment (Kixunil) [reply bitcoin/src/blockdata/transaction.rs:1007]: That is non-bitcoin-y about it? "transaction inputs" is normal in speech, at least where I live. :)

File: bitcoin/src/crypto/sighash.rs
Comment (Kixunil) [inline bitcoin/src/crypto/sighash.rs:1337]: Damn I didn't notice it's still inconsistent because of 0x, but a minor issue, not going to hold up the PR.

**Review by Kixunil — APPROVED** [review]
ACK e356ff661182eb15b9d3d18b2c200f20e3bda76e

**Review by apoelstra — APPROVED** [review]
ACK e356ff661182eb15b9d3d18b2c200f20e3bda76e

File: bitcoin/src/crypto/sighash.rs
Comment (apoelstra) [reply bitcoin/src/crypto/sighash.rs:1337]: Yeah, oops. But agreed, no need to hold up the PR. At least we're not mixing bases.

**Pull request merged**
